tensorflow/python/kernel_tests/conv_ops_3d_test.py

```diff
@@ -80,23 +90,42 @@ def _SetupValuesForDevice(self, tensor_in_sizes, filter_in_sizes, stride,
   def _VerifyValues(self, tensor_in_sizes, filter_in_sizes, stride, padding,
                     expected):
     results = []
-    for data_format, use_gpu in GetTestConfigs():
-      result = self._SetupValuesForDevice(
-          tensor_in_sizes,
-          filter_in_sizes,
-          stride,
-          padding,
-          data_format,
-          use_gpu=use_gpu)
-      results.append(result)
-      tolerance = 1e-2 if use_gpu else 1e-5
+    for (data_format, use_gpu) in GetTestConfigs():
+      for dtype in self._DtypesToTest(use_gpu):
+        result = self._SetupValuesForDevice(
+            tensor_in_sizes,
+            filter_in_sizes,
+            stride,
+            padding,
+            data_format,
+            dtype,
+            use_gpu=use_gpu)
+        results.append((result, use_gpu, dtype, data_format))
+
       with self.test_session() as sess:
-        values = sess.run(results)
-        for value in values:
+        tensor_results = [result[0] for result in results]
+        test_params = [result[1:] for result in results]
+        values = sess.run(tensor_results)
+        for i in range(len(tensor_results)):
+          value = values[i]
+          print("use_gpu:", test_params[i][0])
+          print("dtype:", test_params[i][1])
+          print("data_format:", test_params[i][2])
+
           print("expected = ", expected)
           print("actual = ", value)
-          self.assertAllClose(expected, value.flatten(), atol=tolerance,
-                              rtol=1e-6)
+          tol = 1e-5
+          if value.dtype == np.float16:
+            tol = 1e-3
+            # fp16 using may result in inf values and large absolute errors
+            # when used with large numbers
+            if np.any(value > 1e4):
+              print("fp16 using may result in inf values and large absolute "
+                    "errors when used with large numbers, skipping")
+              continue
+
+          self.assertAllClose(expected, value.flatten(), atol=tol,
+                              rtol=tol)
```

Comment:
Before, fp32 rtol was 1e-6, but now its 1e-5. It should be changed back to 1e-6 for fp16.

Reply:
Ok, thanks for mention.